File: Frontend/src/features/config/solicitudApi.ts
import { api } from '../../lib/apiClient'
import type { Alcance } from './permisoApi'

export const ESTADO_SOLICITUD = {
  pendiente: 0,
  aprobada: 1,
  rechazada: 2,
} as const

export type EstadoSolicitud = (typeof ESTADO_SOLICITUD)[keyof typeof ESTADO_SOLICITUD]

export interface SolicitudPermisoResponse {
  id: number
  usuarioId: number
  /** Nombre de quien pide. */
  usuario: string
  submodulo: string
  accion: string
  motivo: string | null
  /** El documento sobre el que iba, si venia de uno. */
  referencia: string | null
  estado: EstadoSolicitud
  fechaSolicitud: string
  fechaResolucion: string | null
  respuesta: string | null
}

export const solicitudApi = {
  /** POST /api/permiso/solicitar — pedir una acción bloqueada. */
  solicitar: (body: {
    submodulo: string
    accion: string
    motivo?: string | null
    referencia?: string | null
  }) => api.post<SolicitudPermisoResponse>('/permiso/solicitar', body),

  /** GET /api/permiso/solicitudes — la bandeja del admin. */
  bandeja: (soloPendientes = false) =>
    api.get<SolicitudPermisoResponse[]>(`/permiso/solicitudes?soloPendientes=${soloPendientes}`),

  /** GET /api/permiso/solicitudes/mias */
  mias: () => api.get<SolicitudPermisoResponse[]>('/permiso/solicitudes/mias'),

  /** POST /api/permiso/solicitudes/{id}/aprobar */
  aprobar: (id: number, body: { alcance: Alcance; expiraEn?: string | null; respuesta?: string | null }) =>
    api.post<SolicitudPermisoResponse>(`/permiso/solicitudes/${id}/aprobar`, body),

  /** POST /api/permiso/solicitudes/{id}/rechazar */
  rechazar: (id: number, respuesta?: string | null) =>
    api.post<SolicitudPermisoResponse>(`/permiso/solicitudes/${id}/rechazar`, { respuesta }),
}

File: titanic/lib/core/permisos/solicitud_api.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../features/auth/estado/auth_controlador.dart';
import '../red/cliente_api.dart';

/// Pedir un permiso con el que uno se topó bloqueado.
///
/// Existe aparte del resto de llamadas porque no pertenece a ningún módulo:
/// cualquiera puede topar con una acción bloqueada, en cualquier pantalla.
class SolicitudApi {
  const SolicitudApi(this._api);

  final ClienteApi _api;

  /// POST /api/permiso/solicitar
  ///
  /// Pedir dos veces lo mismo no duplica nada: el backend devuelve la
  /// solicitud que ya estuviera abierta.
  Future<void> solicitar({
    required String submodulo,
    required String accion,
    String? motivo,
    String? referencia,
  }) => _api.post(
    '/permiso/solicitar',
    cuerpo: {
      'submodulo': submodulo,
      'accion': accion,
      'motivo': motivo,
      'referencia': referencia,
    },
  );

  /// GET /api/permiso/solicitudes/mias — solo las que siguen sin respuesta,
  /// como claves "submodulo:accion".
  ///
  /// Se filtra aquí y no en la pantalla porque lo resuelto no sirve de nada:
  /// una solicitud aprobada ya se ve como permiso concedido, y una rechazada
  /// se puede volver a pedir.
  Future<Set<String>> misPendientes() async {
    final datos = await _api.get('/permiso/solicitudes/mias') as List;

    return {
      for (final e in datos.cast<Map<String, dynamic>>())
        // 0 es pendiente en el backend.
        if ((e['estado'] as int? ?? 0) == 0) '${e['submodulo']}:${e['accion']}',
    };
  }
}

final solicitudApiProvider = Provider((ref) => SolicitudApi(ref.watch(clienteApiProvider)));
